### PR TITLE
[Api][Checkout] cover skipping payment steps in behat

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 1]
```diff
@@ -169,28 +170,51 @@ public function iCompleteTheAddressingStep(): void
...
-     * @Then I should see the thank you page
+     * @When I complete the shipping step
      */
-    public function iShouldSeeTheThankYouPage(): void
+    public function iCompleteTheShippingStep(): void
     {
-        /** @var Response $response */
-        $response = $this->client->getResponse();
+        /** @var ShippingMethodInterface $shippingMethod */
+        $shippingMethod = $this->shippingMethodRepository->findBy([], null, 1)[0];
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:178]:
```suggestion
        $shippingMethod = $this->shippingMethodRepository->findOneBy([]);
```

File: features/checkout/leaving_notes_on_order_during_checkout.feature [hunk 2]
```diff
@@ -13,7 +13,7 @@ Feature: Leaving additional request notes on my order during checkout
         And there is a customer account "customer@example.com" identified by "sylius"
         And I am logged in as "customer@example.com"
 
-    @ui @api
+    @api
```
Comment (GSadee) [inline features/checkout/leaving_notes_on_order_during_checkout.feature:16]: Should be reverted

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 3]
```diff
@@ -35,21 +37,27 @@ final class CheckoutContext implements Context
...
     ) {
         $this->client = $client;
         $this->responseChecker = $responseChecker;
         $this->sharedStorage = $sharedStorage;
+        $this->shippingMethodRepository = $shippingMethodRepository;
     }
 
     /**
      * @Given I am at the checkout addressing step
+     * @When I complete the shipping step
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:60]: Is it still needed?

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 4]
```diff
@@ -186,29 +195,72 @@ public function iConfirmMyOrder(): void
      */
     public function iCompleteTheAddressingStep(): void
     {
+        if (!array_key_exists('email', $this->content)) {
+             $this->iSpecifyTheEmailAs('customer@example.com');
+        }
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:200]: IMO it should be reverted, as there already is setting customer as null in https://github.com/Sylius/Sylius/blob/master/src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php#L229:L231

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 5]
```diff
@@ -186,29 +195,72 @@ public function iConfirmMyOrder(): void
...
+    }
 
-        Assert::same($value, OrderCheckoutStates::STATE_ADDRESSED);
+    /**
+     * @Then I should be on the checkout complete step
+     * @Then I should be on the checkout summary step
+     */
+    public function iShouldBeOnTheCheckoutCompleteStep(): void
+    {
+        Assert::same($this->getCheckoutState(), OrderCheckoutStates::STATE_PAYMENT_SKIPPED);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:239]: This may not always be true

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 6]
```diff
@@ -235,4 +287,11 @@ private function getHeaders(array $headers = []): array
 
         return $headers;
     }
+    private function getCheckoutState(): string
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:290]: Missing blank line above

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php [hunk 7]
```diff
@@ -69,6 +74,7 @@ public function iTryToOpenCheckoutShippingPage()
 
     /**
      * @When /^I(?:| try to) complete the shipping step$/
+     * @When /^I complete the shipping step with first shipping method$/
      */
     public function iCompleteTheShippingStep()
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php:79]: Missing return type

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php [hunk 8]
```diff
@@ -41,21 +42,25 @@ public function __construct(
...
+     * @Given I have selected :shippingMethod shipping method
+     * @When I select :shippingMethod shipping method
      */
-    public function iSelectShippingMethod(string $shippingMethodName): void
+    public function iSelectShippingMethod(?ShippingMethodInterface $shippingMethod): void
     {
+        $shippingMethodName = 'Free';
+        if ($shippingMethod !== null) {
+            $shippingMethodName = $shippingMethod->getName();
+        }
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php:63]: Why did you add it? Shipping method shouldn't be null

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php [hunk 9]
```diff
@@ -69,6 +74,7 @@ public function iTryToOpenCheckoutShippingPage()
 
     /**
      * @When /^I(?:| try to) complete the shipping step$/
+     * @When /^I complete the shipping step with first shipping method$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php:77]:
```suggestion
     * @When I complete the shipping step with first shipping method
```

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php [hunk 10]
```diff
@@ -41,21 +42,25 @@ public function __construct(
     }
 
     /**
-     * @Given I have proceeded selecting :shippingMethodName shipping method
-     * @When I proceed with :shippingMethodName shipping method
+     * @Given I have proceeded selecting :shippingMethod shipping method
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php:45]: What is the reason you changed the implementation of steps in ui context?

File: src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml [hunk 11]
```diff
@@ -24,10 +24,14 @@ default:
                 - sylius.behat.context.setup.channel
                 - sylius.behat.context.setup.payment
                 - sylius.behat.context.setup.product
+                - sylius.behat.context.setup.promotion
                 - sylius.behat.context.setup.shipping
                 - sylius.behat.context.setup.shop_api_security
 
                 - sylius.behat.context.api.admin.managing_orders
+                - sylius.behat.context.setup.shop_security
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/suites/api/checkout/checkout.yml:32]: Should be removed

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php (hunk 10 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php:45]: at one of the API steps, I needed to get `shippingMethod` instead `shipping method name` that implicated changes also in UI steps

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php (hunk 8 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php:63]: Because i reuse function `iSelectShippingMethod` where `ShippingMethod` can be null because function: `iProceedSelectingBillingCountryAndShippingMethod` can have null as a agument

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php (hunk 10 above)
Comment (GSadee) [reply src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutShippingContext.php:45]: I understand, but this shouldn't force changes in UI context, you can have it implemented `@Given I have proceeded selecting :shippingMethodName shipping method` in UI context and `@Given I have proceeded selecting :shippingMethod shipping method` in API context

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 12]
```diff
@@ -204,24 +214,78 @@ public function iCompleteTheAddressingStep(): void
...
+        Assert::same(
+            $this->responseChecker->getResponseContent($response)['payments'][0]['method']['name'],
+            $paymentMethod->getName()
+        );
+    }
+
+    /**
+     * @Then I should not see any information about payment method
+     */
+    public function iShouldNotSeeAnyInformationAboutPaymentMethod()
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:283]:
```suggestion
    public function iShouldNotSeeAnyInformationAboutPaymentMethod(): void
```

File: src/Sylius/Behat/Context/Ui/Shop/CheckoutContext.php [hunk 13]
```diff
@@ -116,7 +117,7 @@ public function iProceedSelectingPaymentMethod($paymentMethodName)
      * @Given I have proceeded order with :shippingMethodName shipping method and :paymentMethodName payment
      * @When I proceed with :shippingMethodName shipping method and :paymentMethodName payment
      */
-    public function iProceedOrderWithShippingMethodAndPayment($shippingMethodName, $paymentMethodName)
+    public function iProceedOrderWithShippingMethodAndPayment(string $shippingMethodName, $paymentMethodName)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/CheckoutContext.php:120]:
```suggestion
    public function iProceedOrderWithShippingMethodAndPayment(string $shippingMethodName, string $paymentMethodName): void
```

**Comment by GSadee** [comment]
Thank you, Adam! :tada:
